### ISSUE TITLE
how do you inject a prolog/epilog that refers to an external assembly?

### ISSUE THREAD

**Issue by sanikolov** [microsoft/CLRInstrumentationEngine#177]
In callback BeforeInstrumentMethod() I'd like to call pInstructionGraph->CreateBaseline(...) such that either a prolog or an epilog or both get injected as shown below.
```
[... prolog ...] which does 
   call vararg int32 MyOwn.ValidatorClass.ValidateProlog(__arglist)
originalMethodBody here
[...epilog ...]  which does
   call vararg int32 MyOwn.ValidatorClass.ValidateEpilog(__arglist)
ret
```

where ValidatorClass is implemented by me and lives in assembly MyOwn.dll .
See this [old article](https://www.codeproject.com/Articles/4567/The-Implementation-of-Model-Constraints-in-NET-II) where the author makes the following statement 

> Look at the CC.IL module provided as an example in \Barracuda2\CChecker\IL folder to see that it doesn't refer to CCCore.dll. So, we "dynamically link" this dll at runtime.

How do you achieve the same with the CLR Instrumentation Engine?
If Nagler implements this to support one of the many test cases, I couldn't recognize the implementation.

**Comment by wiktork** [comment]
How you do this depends on where MyOwn.dll is located:

1. If MyOwn.dll is loadable from the app you are instrumenting you can call DefineAssemblyRef during ModuleLoad events and create a reference from one assembly to another. IModuleInfo::GetMetadataAssemblyEmit can be used for this. You can also use IMetadataEmit to create the appropriate MemberRef's. Once you get the callback to instrument the method of interest, you can inject a call instruction with your MemberRef token into the instruction graph.

2. If you cannot load MyOwn.dll directly into the load context, things can get a bit tricky. For example, you may need to emit instructions to call Assembly.LoadFrom first and then emit calls to reflection api's to call the desired methods inside the dynamically loaded modules.

Can you elaborate on your scenario a little bit more? Do you own the app and the environment you are trying to instrument?

**Comment by sanikolov** [comment]
Thanks for the quick reply. The app I am instrumenting is 3rd party, thus I have little control over it.
I am looking to instrument a few methods in that app, in a generic way, meaning that the same prolog and epilog validations will be invoked for all of these methods. The validation which resides in MyOwn.dll gets a method token and __arglist, then looks into the method token and walks the __arglist to decides what to do.

**Comment by delmyers** [comment]
Hi @sanikolov,

Unfortunately, you are going to have to work around a few things, then.

1) You will have to get your dll loaded. That probably means that you will have to call an AssemblyLoadFrom as the first step in your prolog. You can add additional guards around it to make sure that you don't have to make the reflection call multiple times.
2) You will have to make sure that you can load all of the assembly's dependencies. If you just depend on standard assemblies, you are probably fine. Otherwise, you will have to do something like create bootstrapper that can respond to assembly load requests by writing a custom assembly resolver.
3) You will have to make sure that the assembly loader will not block your assembly from loading. Sign it, and make sure that, on your target machine, the dll isn't "blocked". This is something that the OS will do whenever you download a dll from the internet. You have to unblock any dll before it can be loaded.

**Comment by sanikolov** [comment]
Isn't the recommendation to use Assembly.LoadFrom(...) equivalent to calling LoadLibraryEx("MyOwn.dll",...) as shown in [NaglerIstrumentationMethod.cpp](https://github.com/microsoft/CLRInstrumentationEngine/blob/63a0979f04d4569b3e84f875101943a5b63a61c2/tests/InstrEngineTests/NaglerInstrumentationMethod/NaglerInstrumentationMethod.cpp#L665) ?

**Comment by SergeyKanzhelev** [comment]
@sanikolov have you seen this instrumentation method: http://apmtips.com/blog/2016/11/18/how-application-insights-status-monitor-not-monitors-dependencies/

This may be exactly what you need. Basically you can insert prolog/epilog to any method by calling a managed method and providing callbacks to be called:

```
Decorator.InitializeExtension();
Functions.Decorate("System", "System.dll", "System.Net.Mail.SmtpClient.Send",
    OnBegin, OnEnd, OnException, false);
```

```
public static object OnBegin(object thisObj, object arg1)
{
    // start the operation
    var operation = new TelemetryClient().StartOperation<DependencyTelemetry>("Send");
    operation.Telemetry.Type = "Smtp";
    operation.Telemetry.Target = ((SmtpClient)thisObj).Host;
    if (arg1 != null)
    {
        operation.Telemetry.Data = ((MailMessage)arg1).Subject;
    }
    // save the operation in the local context
    return operation;
}
```

Is it something you are looking to implement?

**Comment by sanikolov** [comment]
@SergeyKanzhelev yes this looks great. It appears to be built on top of CLRIE which ensures coexistence with other APM/Intellitrace apps, right? Functions.Decorate() appears to come from the mysterious Microsoft.Diagnostics.Instrumentation.Extensions.Intercept package which is marked as prerelease and unlisted in nuget, has a project website on nuget that sheds very little light on it, and there is no repo for it that I could find, despite its MIT license. If it is closed source perhaps we won't be able to use it. But overall it looks great.

**Comment by delmyers** [comment]
@SergeyKanzhelev,

If I understand the apm status monitor, it requires you to install the App Insights SDK as part of the application, correct? Doesn't it depend on the callback dll being discoverable by the application?

@sanikolov, unfortunately, the code that you linked to is only there to allow the NaglerProfiler to load the given dll into memory so that it can read IL. That doesn't allow it to actually execute any of the code. The CLR has to do additional work after the dll is loaded into memory before it can actually execute code. That is why "LoadFrom" must be called from managed code.

**Comment by sanikolov** [comment]
Hello again. I was hoping to extract the operand bytes via the API provided in [InstrumentationEngine.h#L1669](https://github.com/microsoft/CLRInstrumentationEngine/blob/e78064105500b63618b24d05f949613473a708f0/src/InstrumentationEngine.Api/InstrumentationEngine.h#L1669). Alas, I couldn't find an API call to do so.
Is this oversight deliberate, accidental or there is already a roundabout way to extract the operand bytes?

**Comment by jander-msft** [comment]
In the future, please open separate issue since it is not related to prolog/epilog question.

But here's how you can get the information: QI the IInstruction for IOperandInstruction, use IInstruction::GetOperandLength to get its length, and use IOperandInstruction::GetOperandValue to get the byte array for the operand value.
